src/application_default_credentials.rs

```diff
@@ -0,0 +1,48 @@
+use crate::error::Error;
+use crate::types::TokenInfo;
+
+/// Provide options for the Application Default Credential Flow, mostly used for testing
+pub struct ApplicationDefaultCredentialsFlowOpts {
+    /// Used as base to build the url during token request from GCP metadata server
+    pub metadata_url: Option<String>,
+}
+impl Default for ApplicationDefaultCredentialsFlowOpts {
+    fn default() -> Self {
+        Self { metadata_url: None }
+    }
+}
+
+pub struct ApplicationDefaultCredentialsFlow {
+    metadata_url: String,
+}
+
+impl ApplicationDefaultCredentialsFlow {
+    pub(crate) fn new(opts: ApplicationDefaultCredentialsFlowOpts) -> Self {
+        let metadata_url = opts.metadata_url.unwrap_or_else(|| "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token".to_string());
```

Comment:
Is this bias towards Google consistent with the rest of the crate?

Reply:
Application Default Credentials is a flow used in the GCP ecosystem. Thus this is biased towards Google, but not unduly so. None of the other cloud vendors for example implement exactly this type of authentication flow.

for more information see: https://cloud.google.com/docs/authentication/production

Reply:
Thanks @braincow - that's helpful. (I don't seem to be able to close some of my comment threads in this PR, but I would if I could!)